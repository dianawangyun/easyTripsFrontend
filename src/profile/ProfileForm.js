import React, { useState, useContext } from "react";
import EasyTripsApi from "../api/api";
import UserContext from "../auth/UserContext";

/* The form for updating the profile. */
function ProfileForm() {
    const { currentUser, setCurrentUser } = useContext(UserContext);
    const [formData, setFormData] = useState({
        username: currentUser.username,
        firstName: currentUser.firstName,
        lastName: currentUser.lastName,
        email: currentUser.email,
        password: "",
    });
    const [saved, setSaved] = useState(false);
    const [formErrors, setFormErrors] = useState([]);

    async function handleSubmit(e) {
        e.preventDefault();
        let updateData = {
            firstName: formData.firstName,
            lastName: formData.lastName,
            email: formData.email,
            password: formData.password,
        };

        let newProfile;
        try {
            newProfile = await EasyTripsApi.updateProfile(
                currentUser.username,
                updateData
            );
            setFormErrors([]);
            setCurrentUser(newProfile);
            setSaved(true);
        } catch (e) {
            setFormErrors(e);
        }

        setFormData((data) => ({ ...data, password: "" }));
    }

    function handleChange(e) {
        const { name, value } = e.target;
        setFormData((data) => ({ ...data, [name]: value }));
        setSaved(false);
    }
    return (
        <div className="container">
            <h2 className="mt-5 mb-4 text-center">User Profile</h2>
            <div className="row d-flex justify-content-center">
                <div className="col-md-6">
                    <form onSubmit={handleSubmit}>
                        <div className="form-group">
                            <label htmlFor="username">Username:</label>
                            <input
                                className="form-control"
                                type="text"
                                name="username"
                                id="username"
                                value={formData.username}
                                disabled
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="password">Password:</label>
                            <input
                                className="form-control"
                                type="password"
                                name="password"
                                id="password"
                                value={formData.password}
                                onChange={handleChange}
                                required
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="firstname">First Name:</label>
                            <input
                                className="form-control"
                                type="text"
                                name="firstName"
                                id="firstname"
                                value={formData.firstName}
                                onChange={handleChange}
                                required
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="lastname">Last Name:</label>
                            <input
                                className="form-control"
                                type="text"
                                name="lastName"
                                id="lastname"
                                value={formData.lastName}
                                onChange={handleChange}
                                required
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="email">Email:</label>
                            <input
                                className="form-control"
                                type="text"
                                name="email"
                                id="email"
                                value={formData.email}
                                onChange={handleChange}
                                required
                            />
                        </div>
                        {formErrors.length ? (
                            <ul>
                                {formErrors.map((e) => (
                                    <li className="text-danger">{e}</li>
                                ))}
                            </ul>
                        ) : null}
                        {saved ? (
                            <p className="text-success">Profile updated.</p>
                        ) : null}
                        <div className="text-center">
                            <button className="btn btn-success" type="submit">
                                Save
                            </button>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    );
}

export default ProfileForm;
