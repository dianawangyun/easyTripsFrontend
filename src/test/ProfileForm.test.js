import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ProfileForm from "../profile/ProfileForm";

import UserContext from "../auth/UserContext";

function renderProfileForm(currentUser, setCurrentUser) {
    return render(
        <UserContext.Provider value={{ currentUser, setCurrentUser }}>
            <ProfileForm />
        </UserContext.Provider>
    );
}

const currentUser = {
    username: "testUsername",
    firstName: "test",
    lastName: "test",
    email: "email",
};
const setCurrentUser = () => {
    return;
};

it("renders without crashing", function () {
    renderProfileForm(currentUser, setCurrentUser);
});

it("matches snapshot", function () {
    const { asFragment } = renderProfileForm(currentUser, setCurrentUser);
    expect(asFragment()).toMatchSnapshot();
});

it("shows the form with data", function () {
    const { getByLabelText } = renderProfileForm(currentUser, setCurrentUser);
    const userName = getByLabelText("Username:");
    expect(userName.value).toEqual("testUsername");
});
