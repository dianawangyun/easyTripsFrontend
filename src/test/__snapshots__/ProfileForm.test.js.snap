// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h2
      class="mt-5 mb-4 text-center"
    >
      User Profile
    </h2>
    <div
      class="row d-flex justify-content-center"
    >
      <div
        class="col-md-6"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              for="username"
            >
              Username:
            </label>
            <input
              class="form-control"
              disabled=""
              id="username"
              name="username"
              type="text"
              value="testUsername"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password:
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              required=""
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="firstname"
            >
              First Name:
            </label>
            <input
              class="form-control"
              id="firstname"
              name="firstName"
              required=""
              type="text"
              value="test"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="lastname"
            >
              Last Name:
            </label>
            <input
              class="form-control"
              id="lastname"
              name="lastName"
              required=""
              type="text"
              value="test"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="email"
            >
              Email:
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              required=""
              type="text"
              value="email"
            />
          </div>
          <div
            class="text-center"
          >
            <button
              class="btn btn-success"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
